### PR TITLE
Fix formatting of tables where headers span multiple rows

### LINKED ISSUE #6482: $PSVersionTable.PSVersion throws error System.ArgumentOutOfRangeException,Microsoft.PowerShell.Commands.OutLineOutputCommand
$PSVersionTable.PSVersion.ToString() and $PSVersionTable still works.

The error can be reproduced on master branch.

Steps to reproduce
------------------

```powershell

$PSVersionTable.PSVersion
out-lineoutput : Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
+ CategoryInfo          : NotSpecified: (:) [out-lineoutput], ArgumentOutOfRangeException
+ FullyQualifiedErrorId : System.ArgumentOutOfRangeException,Microsoft.PowerShell.Commands.OutLineOutputCommand

```

Expected behavior
-----------------

```none
Get the PSVersion string
```

Actual behavior
---------------
Error - System.ArgumentOutOfRangeException,Microsoft.PowerShell.Commands.OutLineOutputCommand

```none

```

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.1
PSEdition                      Core
GitCommitId                    v6.0.1-290-g942832c13a3c325851e9873bd25036775c85a18c
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

This was originally found where $PSversionTable.PSVersion was broken, but the real problem is whenever a table header spans multiple rows due to limited width, the calculation was incorrect for ensuring
no trailing whitespace.

To do this correctly, we need to calculate the last column that has content and add empty strings for cells that are empty.  Also make sure to trim the trailing whitespace from the last column that has content for
a given row.

Fix https://github.com/PowerShell/PowerShell/issues/6482

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 1]
```diff
@@ -289,22 +289,64 @@ private string[] GenerateTableRow(string[] values, int[] alignment, DisplayCells
...
+            //
+            // Long Header2 Head
+            // Head         er3
+            // er
+            // ---- ------- ----
+            // 1    2       3
+            //
+            // To ensure we don't add whitespace to the end, we need to determine the last column in each row with content
+
+            int[] lastColWithContent = new int[screenRows];
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: We could use new C# Span:
```c#
Span<int> lastColWithContent = stackalloc int[screenRows]; 
```

**Comment by SteveL-MSFT** [comment]
@iSazonov you should still have `merge` permission.  We changed a setting on `master` branch so that an approved review is required before `merge` is available

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: good suggestion, will change

**Comment by iSazonov** [comment]
@SteveL-MSFT #6490 was approved but I haven't permissions.

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: [System.Memory](https://www.nuget.org/packages/System.Memory/4.5.0-preview1-26216-02) is still preview, though

**Comment by daxian-dbw** [comment]
@iSazonov I just added you to the "people and teams" allowed to push to the master branch. Can you please try again to see if you can merge now?

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: I see it is compatible with .NETCoreApp 2.0.

Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: It's compatible, but do we want to take dependency on something that's Preview?

Comment (iSazonov) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: Seems we have plans to move .Net Core 2.1.0 Preview in days/weeks. If so we can start here and now.

Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: If we expect System.Memory to GA when .Net Core 2.1 GA's I think this is fine.  We can revert this line later if this turns out to not be true.

Comment (iSazonov) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: https://blogs.msdn.microsoft.com/dotnet/2018/02/02/net-core-2-1-roadmap/

Span<T> is in .Net Core 2.1 roadmap.

File: src/System.Management.Automation/System.Management.Automation.csproj [hunk 2]
```diff
@@ -4,6 +4,7 @@
     <Description>PowerShell Core's System.Management.Automation project</Description>
     <NoWarn>$(NoWarn);CS1570;CS1734</NoWarn>
     <AssemblyName>System.Management.Automation</AssemblyName>
+    <LangVersion>7.2</LangVersion>
```
Comment (iSazonov) [inline src/System.Management.Automation/System.Management.Automation.csproj]: Do we really need this? Currently we use default and seems the default is 7.2.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 [hunk 3]
```diff
@@ -339,4 +339,191 @@ Left Center Right
...
+		</View>
+	</ViewDefinitions>
+</Configuration>
+"@; expectedTable = @"
+
+Long Header2 Head
+Long         er3
+Head
+er
+---- ------- ----
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: I believe we should have a test for edge case - one column.

File: src/System.Management.Automation/System.Management.Automation.csproj (hunk 2 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/System.Management.Automation.csproj]: Span was not found and from what I could find, currently we need to opt into 7.2 features.  Perhaps it will be default once dotnetcore2.1 GAs.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 (hunk 3 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: Will add

Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: I'm also going to turn the ps1xml into a template since most of it is the same

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 [hunk 4]
```diff
@@ -339,4 +339,141 @@ Left Center Right
...
+er
+----
+1
+
+
+
+"@ }
+		) {
+			param($view, $widths, $expectedTable)
+			$ps1xml = @"
```
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: NIT: This is rather hard to read with large here strings in the middle of the code.

**Review by TravisEz13 — COMMENTED** [review]
reviewed with one minor comment

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 (hunk 4 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: In VSCode, the coloring makes it much easier to read.  Collapsing the XML will make it harder to read the XML.

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: I agree too that it is more editors and formatting plugins issue and it strongly depends on personal preferences. We could be more happy with the support of YAML.

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs [hunk 5]
```diff
@@ -314,10 +356,18 @@ private string[] GenerateTableRow(string[] values, int[] alignment, DisplayCells
             for (int row = 0; row < rows.Length; row++)
             {
                 StringBuilder sb = new StringBuilder();
-                // for a give row, walk the columns
+                // for a given row, walk the columns
                 for (int col = 0; col < scArray.Length; col++)
```
Comment (iSazonov) [inline src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: Not for th PR - we have several similar repetitive cycles which we could replace one.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 [hunk 6]
```diff
@@ -339,4 +339,141 @@ Left Center Right
...
+					</TableRowEntry>
+				</TableRowEntries>
+			</TableControl>
+		</View>
+	</ViewDefinitions>
+</Configuration>
+"@
+			$ps1xml = $ps1xml.Replace("{0}", $widths[0]).Replace("{1}", $widths[1]).Replace("{2}", $widths[2])
+			$ps1xmlPath = Join-Path -Path $TestDrive -ChildPath "test.format.ps1xml"
+			Set-Content -Path $ps1xmlPath -Value $ps1xml
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: Maybe add `-Force`?

**Review by iSazonov — APPROVED** [review]
With minor comments

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 5 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: Please open an issue and mark as code cleanup.  Thanks!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1 (hunk 6 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Format-Table.Tests.ps1]: Isn't needed here, `-Force` is only needed if the file is read-only

File: src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/FormatAndOutput/common/TableWriter.cs]: Done #6526.

**Comment by iSazonov** [comment]
@TravisEz13 Can you sign off?

File: src/System.Management.Automation/System.Management.Automation.csproj (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj]: @SteveL-MSFT I found in [docs](https://docs.microsoft.com/en-us/dotnet/csharp/reference-semantics-with-value-types) that we could use `Latest` - if this works maybe we expand this on all proj files?

/cc @daxian-dbw

Comment (daxian-dbw) [reply src/System.Management.Automation/System.Management.Automation.csproj]: The `<LanguageVersion>` tag should be put in `PowerShell.Common.props`, I think.

**Review by TravisEz13 — APPROVED** [review]
I forced the package changes.  I'm good if they pass

File: src/System.Management.Automation/System.Management.Automation.csproj (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/System.Management.Automation.csproj]: I opened the tracking Issue #6547.
